### PR TITLE
Optimize certificate cache

### PR DESCRIPTION
**Pull request by moticless**
#### What does this PR do?
Adding cache mechanism for provider certificates that will refresh them from provider only when client fails to decode with cached certificate.

Fix cucumber tests - fix tests to lookup for correct errors in log

#### Any background context you want to provide?
OIDC authentication verifies given TokenID by retrieving certificate from OIDC provider and verifying token signature. Following this feature it won't retrieve certificate per each request but cache it. A new retrieval will be triggered only on failure.

#### What ticket does this PR close?
Connected to #927 

#### Where should the reviewer start?
read relevant section at :  https://ca-il-confluence.il.cyber-ark.com/display/rndp/Research++-+rails+caching+mechanism+and+cert+optimization#?lucidIFH-viewer-599ae143=1

Then start with certificate_cache.rb - it stands on its own.

#### How should this be manually tested?
- Run several oidc requests and verify only one time provider is accessed to retrieve certificate.
- Rotate provider certificate and verify OIDC handler knows how to recover and retrieve new certificate.

### CODE REVIEW

File: app/domain/authentication/authn_oidc/authenticate_id_token/fetch_provider_certificate.rb [hunk 1]
```diff
@@ -12,21 +12,29 @@ module AuthenticateIdToken
 
       FetchProviderCertificate = CommandClass.new(
         dependencies: {
-          logger: Rails.logger
+          logger: Rails.logger,
+          cert_cache: lambda {::Authentication::AuthnOidc::CertificateCache.instance}
```
Comment (orenbm) [inline app/domain/authentication/authn_oidc/authenticate_id_token/fetch_provider_certificate.rb]: @jonahx is this ok?

**Comment by nessiLahav** [comment]
@moticless Please do the following:
1. add logs
2. also test it with Okta
3. crate a new exception on X failures for the last Y hours, because it should be a rare situation  (maybe an attacker or multiple mistakes in a row)

File: app/domain/authentication/authn_oidc/authenticate_id_token/fetch_provider_certificate.rb [hunk 2]
```diff
@@ -8,28 +8,27 @@ module AuthenticateIdToken
...
         dependencies: {
-          logger: Rails.logger
+          logger: Rails.logger,
+          open_id_discovery_service: OpenIDConnect::Discovery::Provider::Config
         },
         inputs: %i(provider_uri)
       ) do
 
         def call
-          @logger.debug(Log::OIDCProviderUri.new(@provider_uri).to_s)
```
Comment (orenbm) [inline app/domain/authentication/authn_oidc/authenticate_id_token/fetch_provider_certificate.rb]: was this removed intentionally? i re-added it

Comment (orenbm) [reply app/domain/authentication/authn_oidc/authenticate_id_token/fetch_provider_certificate.rb]: added it like this so we won't have code in `call` method:
```
        def call
          log_provider_uri
          discover_provider
          fetch_certs
        end

        private

        def log_provider_uri
          @logger.debug(Log::OIDCProviderUri.new(@provider_uri).to_s)
        end
```

File: app/domain/authentication/authn_oidc/authenticate_id_token/fetch_provider_certificate.rb [hunk 3]
```diff
@@ -38,9 +37,9 @@ def discover_provider
         end
 
         def fetch_certs
-          jwks = @discovered_provider.jwks
-          @logger.debug(Log::FetchProviderCertsSuccess.new.to_s)
-          jwks
+          @discovered_provider.jwks.tap do
+            @logger.debug(Log::FetchProviderCertsSuccess.new.to_s)
+          end
```
Comment (orenbm) [inline app/domain/authentication/authn_oidc/authenticate_id_token/fetch_provider_certificate.rb]: that's nice

**Pull request merged**